drv/i2c-devices/src/mwocp68.rs

```diff
@@ -350,6 +440,262 @@ impl Mwocp68 {
+            let mut data = [0u8; MWOCP68_BOOT_LOADER_KEY.len() + 2];
+
+            data[0] = CommandCode::BOOT_LOADER_KEY as u8;
+            data[1] = MWOCP68_BOOT_LOADER_KEY.len() as u8;
+            data[2..].copy_from_slice(MWOCP68_BOOT_LOADER_KEY);
+
+            self.device
+                .write(&data)
+                .map_err(|code| Error::BadWrite { cmd: data[0], code })?;
+
+            Ok(UpdateState::WroteBootLoaderKey)
+        };
+
+        let write_product_key = || -> Result<UpdateState, Error> {
+            const MWOCP68_PRODUCT_KEY: &[u8] = b"M5813-0000000000";
+            let mut data = [0u8; MWOCP68_PRODUCT_KEY.len() + 1];
+
+            data[0] = CommandCode::BOOT_LOADER_PRODUCT_KEY as u8;
+            data[1..].copy_from_slice(MWOCP68_PRODUCT_KEY);
+
+            self.device
+                .write(&data)
+                .map_err(|code| Error::BadWrite { cmd: data[0], code })?;
+
+            Ok(UpdateState::WroteProductKey)
+        };
+
+        let boot_boot_loader = || -> Result<UpdateState, Error> {
+            self.boot_loader_command(BootLoaderCommand::BootPrimary)?;
+            Ok(UpdateState::BootedBootLoader)
+        };
+
+        let start_programming = || -> Result<UpdateState, Error> {
+            self.boot_loader_command(BootLoaderCommand::RestartProgramming)?;
+            Ok(UpdateState::StartedProgramming)
+        };
+
+        let write_block = || -> Result<UpdateState, Error> {
+            const len: usize = MWOCP68_BLOCK_LENGTH;
+
+            let (mut offset, mut checksum) = match state {
+                Some(UpdateState::WroteBlock { offset, checksum }) => {
+                    (offset, checksum)
+                }
+                Some(UpdateState::StartedProgramming) => (0, 0),
+                _ => panic!(),
+            };
+
+            let mut data = [0u8; len + 1];
+            data[0] = CommandCode::BOOT_LOADER_MEMORY_BLOCK as u8;
+            data[1..].copy_from_slice(&payload[offset..offset + len]);
+
+            self.device
+                .write(&data)
+                .map_err(|code| Error::BadWrite { cmd: data[0], code })?;
+
+            checksum = data[1..].iter().fold(checksum, |c, &d| c + d as u64);
+            offset += MWOCP68_BLOCK_LENGTH;
+
+            if offset >= payload.len() {
+                Ok(UpdateState::WroteLastBlock { checksum })
+            } else {
+                Ok(UpdateState::WroteBlock { offset, checksum })
+            }
+        };
+
+        let send_checksum = || -> Result<UpdateState, Error> {
+            let mut data = [0u8; 4];
+
+            let checksum = match state {
+                Some(UpdateState::WroteLastBlock { checksum }) => checksum,
+                _ => panic!(),
+            };
+
+            data[0] = CommandCode::IMAGE_CHECKSUM as u8;
+            data[1] = 2;
+            data[2] = (checksum & 0xff) as u8;
+            data[3] = ((checksum >> 8) & 0xff) as u8;
```

Comment:
take it or leave it, but why not:
```suggestion
            let checksum = match state {
                Some(UpdateState::WroteLastBlock { checksum }) => checksum,
                _ => panic!(),
            };
            
            let data: [u8; 4] = [
                CommandCode::IMAGE_CHECKSUM as u8,
                2,
                (checksum & 0xff) as u8,
                ((checksum >> 8) & 0xff) as u8,
            ];
```

Reply:
I tried this on, but it left it out of step with the way the other commands are structured.

Reply:
Actually, I changed my mind: I like it better your way. ;)